src/backend/dx12/src/device.rs

```diff
@@ -1,82 +1,1144 @@
+            d::ResourceHeapType::Any => winapi::D3D12_HEAP_FLAG_ALLOW_ALL_BUFFERS_AND_TEXTURES,
+            d::ResourceHeapType::Buffers => winapi::D3D12_HEAP_FLAG_ALLOW_ONLY_BUFFERS,
+            d::ResourceHeapType::Images  => winapi::D3D12_HEAP_FLAG_ALLOW_ONLY_NON_RT_DS_TEXTURES,
+            d::ResourceHeapType::Targets => winapi::D3D12_HEAP_FLAG_ALLOW_ONLY_RT_DS_TEXTURES,
+        };
+
+        let desc = winapi::D3D12_HEAP_DESC {
+            SizeInBytes: size,
+            Properties: conv::map_heap_properties(heap_type.properties),
+            Alignment: 0, //Warning: has to be 4K for MSAA targets
+            Flags: flags,
+        };
+
+        let hr = unsafe {
+            self.device.CreateHeap(&desc, &dxguid::IID_ID3D12Heap, &mut heap)
+        };
+        if hr == winapi::E_OUTOFMEMORY {
+            return Err(d::ResourceHeapError::OutOfMemory);
+        }
+        assert_eq!(winapi::S_OK, hr);
 
-    fn create_buffer_raw(&mut self, _: buffer::Info) -> Result<handle::RawBuffer<B>, buffer::CreationError> { unimplemented!() }
-    fn create_buffer_immutable_raw(&mut self, data: &[u8], stride: usize, _: buffer::Role, _: Bind)
-                                   -> Result<handle::RawBuffer<B>, buffer::CreationError> { unimplemented!() }
+        //TODO: merge with `map_heap_properties`
+        let default_state = if !heap_type.properties.contains(memory::CPU_VISIBLE) {
+            winapi::D3D12_RESOURCE_STATE_COMMON
+        } else if heap_type.properties.contains(memory::COHERENT) {
+            winapi::D3D12_RESOURCE_STATE_GENERIC_READ
+        } else {
+            winapi::D3D12_RESOURCE_STATE_COPY_DEST
+        };
+
+        Ok(n::Heap {
+            raw: unsafe { ComPtr::new(heap as _) },
+            ty: heap_type.clone(),
+            size,
+            default_state,
+        })
+    }
 
     fn create_renderpass(
         &mut self,
         attachments: &[pass::Attachment],
         subpasses: &[pass::SubpassDesc],
-        dependencies: &[pass::SubpassDependency]
-    ) -> handle::RenderPass<B> {
+        _dependencies: &[pass::SubpassDependency],
+    ) -> n::RenderPass {
+        // TODO:
+        let subpasses = subpasses
+            .iter()
+            .map(|subpass| {
+                n::SubpassDesc {
+                    color_attachments: subpass.color_attachments.iter().cloned().collect(),
+                }
+            }).collect();
+
+        n::RenderPass {
+            attachments: attachments.to_vec(),
+            subpasses,
+        }
+    }
+
+    fn create_pipeline_layout(&mut self, sets: &[&n::DescriptorSetLayout]) -> n::PipelineLayout {
+        // Pipeline layouts are implemented as RootSignature for D3D12.
+        //
+        // Each descriptor set layout will be one table entry of the root signature.
+        // We have the additional restriction that SRV/CBV/UAV and samplers need to be
+        // separated, so each set layout will actually occupy 2 entries!
```

Comment:
could we have a smarter logic that only splits a set layout if it has both samplers and non-samplers?

Reply:
It basically works this way, no empty tables. Comment doesn't reflect it properly (: